### ISSUE TITLE
Docs: tighten security rules

### ISSUE THREAD

**Issue by czabaj** [gavinsawyer/firebase-web-authn#6]
I think that as a part of the extension setup, we should instruct the users to add security rules for `webAuthnUsers` firestore collection, either ban all access from the client

```
match /webAuthnUsers/{userID} {
  allow read, write: if false;
}
```

or allow get one document for the signed-in user, this should be safe and allows a simple check of `lastVerified`

```
match /webAuthnUsers/{userID} {
  allow get: if request.auth.uid == userID;
  allow list: if false;
  allow write: if false;
}
```

Is it possible to apply security rules via Firebase extension?

**Comment by czabaj** [comment]
This is unfortunate - setting up the security rules breaks the functions, it is then no-longer possible to sign-in 🤷

**Comment by gavinsawyer** [comment]
Client access is disallowed by default, so adding the rule isn't necessary. This collection isn't intended to be accessed from the client for security reasons. You would check `lastVerified` in something like an API endpoint to send money. If you did the same logic client-side instead of checking `lastVerified` in the secure context, the user could exploit the endpoint by reverse-engineering the app.

edit: If your app needs to show this, allowing users to read their documents in `webAuthnUsers` is the most performant way to do so. To be clear, though, it's purely presentational and can't be trusted for security logic.

**Comment by gavinsawyer** [comment]
To be sure here's a test from the Firebase Console:

<img width="1180" src="https://github.com/gavinsawyer/firebase-web-authn/assets/6238591/ae92bd6a-1527-474c-aa91-7a278e884718">

**Comment by czabaj** [comment]
> Client access is disallowed by default

@gavinsawyer how could it be? Is there any other mechanism than the security rules involved? In my setup, the `webAuthnUsers` collection is readable from the client without any hassle.

![Screenshot 2023-08-10 at 9 27 27](https://github.com/gavinsawyer/firebase-web-authn/assets/47554112/7ba0d604-0e86-4787-85df-c68a7d10a46d)

And I see the security rules applied in your test, try to remove them.

My point is that we should instruct the users to secure their `webAuthnUsers` collection in their security rules. From what I read, it is impossible to change the security rules from within the extension, users have to make that manually.

**Comment by gavinsawyer** [comment]
It shouldn't be the case that you can access that collection from the client with no rule to allow it. Access is denied unless explicitly allowed except for the Firebase Admin SDK which bypasses all rules. `@firebase-web-authn/server` uses the Admin SDK which is why it can retrieve `lastVerified`, etc. without security rules.

Is it possible that data leaked from your server? Never seen ReScript if I'm being honest so I can't tell where your project is using the Firebase JS SDK.

**Comment by czabaj** [comment]
> Access is denied unless explicitly allowed except for the Firebase Admin SDK which bypasses all rules

It seems it's the case, I just removed my security rules and tried to read document or list documents and the request failed with insufficient permissions.

What mechanism guards the data? I was thinking that only security rules affect what is readable.

**Comment by gavinsawyer** [comment]
> What mechanism guards the data? I was thinking that only security rules affect what is readable.

I had to familiarize myself a bit with IAM in Google Cloud Platform when making this. Interacting with GCP with privilege requires the use of a service account, and the one generated for this extension is given the `datastore.user` role. This means it can read and write on any collection in Firestore, but it's automatically guarded so only the extension's Cloud Functions are using that service account.

Your app's backend for Firebase Hosting is given a generated service account upon `firebase init hosting` as well, named "App Engine default service account." It has the `editor` role which has over 7000 permissions, so almost any requests your app makes server-side will be permitted by GCP. This includes Firestore and the rules for users don't apply here.

**Comment by czabaj** [comment]
Interesting. But I meant the restriction of the Firestore `webAuthnUsers` collection. By default, the client SDK can access "everything anywhere" in the Firestore unless restricted by [the security rules](https://firebase.google.com/docs/firestore/security/get-started), right? But the `webAuthnCollection` is not accessible by default and I cannot find why. Is there some rule, e.g. if you create the collection from the Firebase function it cannot be queried by the client SDK? I just quickly googled for some answers but without luck.

**Comment by gavinsawyer** [comment]
> By default, the client SDK can access "everything anywhere" in the Firestore unless restricted by the security rules, right?

No, by default nothing can be accessed. The CLI is somewhat misleading and says that there's a "test mode" and "locked mode," but these merely write the rules "allow read, write if true;" or "allow read, write if false;" respectively upon initializing the Firestore Database. 

I say it's misleading because the second rule doesn't do anything since collections with no rules are inaccessible to the client by default. This isn't really communicated anywhere but it's apparent from testing.

**Comment by czabaj** [comment]
Oh, you are right! My bad. I started developing the app with `allow read, write if true;` which I think is the default to smoothen the initial development, but the Firebase then forces you to tighten the rules, so I made the simplest thing - `allow read, write if request.auth != null;`, this means that the signed-in user _can_ read "everything anywhere" but that is because I made that rule explicitly 🤦 

I lived with this setup for so long that I got used to it and forgot that I wrote that security rule. By default, no one can read anything with the exception that admin SDK bypasses the security rules 👍
